discovery/sync_manager.go

```diff
@@ -0,0 +1,651 @@
+		err := m.transitionSyncer(s, HistoricalSync)
+		if err == nil {
+			break
+		}
+
+		log.Errorf("Unable to transition GossipSyncer(%x) to %v: %v",
+			s.cfg.peerPub, HistoricalSync, err)
+
+		s = m.chooseRandomSyncer(candidatesChosen)
+	}
+}
+
+// chooseRandomSyncer returns a random non-active syncer that's eligible for a
+// sync transition. A blacklist can be used to skip any previously chosen
+// candidates.
+//
+// NOTE: It's possible for a nil value to be returned if there are no eligible
+// candidate syncers.
+//
+// NOTE: This method requires the syncersMtx read lock to be held.
+func (m *SyncManager) chooseRandomSyncer(
+	blacklist map[routing.Vertex]struct{}) *GossipSyncer {
+
+	for _, s := range m.inactiveSyncers {
+		// Skip any syncers that exist within the blacklist.
+		if blacklist != nil {
+			if _, ok := blacklist[s.cfg.peerPub]; ok {
+				continue
+			}
+		}
+
+		// Only syncers in a chansSynced state are viable for sync
+		// transitions, so skip any that aren't.
+		if s.syncState() != chansSynced {
+			continue
+		}
+
+		return s
+	}
+	return nil
+}
+
+// InitSyncState is called by outside sub-systems when a connection is
+// established to a new peer that understands how to perform channel range
+// queries. We'll allocate a new gossip syncer for it, and start any goroutines
+// needed to handle new queries.
+//
+// TODO(wilmer): Only mark as ActiveSync if this isn't a channel peer.
+func (m *SyncManager) InitSyncState(peer lnpeer.Peer) {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	// If we already have a syncer, then we'll exit early as we don't want
+	// to override it.
+	nodeID := routing.Vertex(peer.PubKey())
+	if _, ok := m.gossipSyncer(nodeID); ok {
+		return
+	}
+
+	log.Infof("Creating new GossipSyncer for peer=%x", nodeID[:])
+
+	encoding := lnwire.EncodingSortedPlain
+	s := newGossipSyncer(GossipSyncerCfg{
+		chainHash:     m.cfg.ChainHash,
+		peerPub:       nodeID,
+		channelSeries: m.cfg.ChanSeries,
+		encodingType:  encoding,
+		chunkSize:     encodingTypeToChunkSize[encoding],
+		sendToPeer: func(msgs ...lnwire.Message) error {
+			return peer.SendMessage(false, msgs...)
+		},
+	})
+
+	// If we don't have any active syncers, we'll default to forcing a full
+	// sync with the first peer we connect to, to ensure we get as much of
+	// the graph as possible. This is mostly done for light clients, which
+	// shouldn't be receiving new updates anyway due to the high bandwidth
+	// consumption that stems from it.
+	if m.cfg.NumActiveSyncers == 0 && len(m.inactiveSyncers) == 0 {
+		s.setSyncType(HistoricalSync)
+		m.inactiveSyncers[nodeID] = s
+		s.Start()
+		return
+	}
+
+	// If we've yet to reach our desired number of active syncers, then
+	// we'll use this one.
+	if len(m.activeSyncers) < m.cfg.NumActiveSyncers {
+		s.setSyncType(ActiveSync)
+		m.activeSyncers[nodeID] = s
+		m.enqueueActiveSyncer(s, false)
+		return
+	}
+
+	// Otherwise, the syncer should be initialized as passive.
```

Comment:
I think that what make this complex is the interaction between the syncer's state and the `syncerHandler`. Here we  modify the syncer state, start it an also modify the `SyncManager`s maps. I'm in favor of giving this responsibility completely to the `syncerHandler`, to avoid weird concurrency behavior.

In `InitSyncState` you would just send the new syncer to the `syncerHandler`, and it would decide which state to start it in. Would also trivially remove the `transition` boolean.

Reply:
Perhaps `syncerHandler` is not the right name for the method, which is causing some confusion? Its main purpose is just to handle the round-robin queue of active syncers, so it could be renamed to `activeSyncerHandler` and the tickers could be moved into a separate handler.